backend/ddang/src/test/java/com/ddang/ddang/notification/application/FcmNotificationServiceTest.java

```diff
@@ -83,10 +87,11 @@ class FcmNotificationServiceTest {
                 "image.png"
         );
 
-        // when & then
-        assertThatThrownBy(() -> notificationService.send(createNotificationDto))
-                .isInstanceOf(DeviceTokenNotFoundException.class)
-                .hasMessage("사용자의 기기 토큰을 찾을 수 없습니다.");
+        // when
+        final String actual = notificationService.send(createNotificationDto);
+
+        // then
+        assertThat(actual).isEqualTo("알림 전송에 실패했습니다.");
```

Comment:
해당 파트가 적어주셨던 `트랜잭션과 테스트 관련 트러블 슈팅` 파트 같네요 

이거는 `@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)`과 `@Transactional`이 같이 묶여서 발생한 이슈라고 생각됩니다 
`@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)`로 인해서 `given`절로 데이터를 넣어주는 트랜잭션과 `then`절에서 조회하는 트랜잭션이 서로 달라서(별도의 스레드) 발생하는 것 같네요 
`given`절에서 저장하기 전(트랜잭션이 끝나기 전)에 `then`절에서 조회하는 스레드의 트랜잭션이 조회하면서 값이 없다고 떠서 발생한 것 같습니다 

그렇다면 `then`절에서 스레드가 트랜잭션에 참여하면서 데이터를 조회하는 시점을 `given`절을 수행하는 스레드가 트랜잭션에 참여해서 데이터를 저장하고 트랜잭션을 커밋하기 전까지 기다리게 하면 되지 않을까요?

```suggestion
TransactionSynchronizationManager.registerSynchronization(new TransactionSynchronization() {
    @Override
    public void afterCommit() {
        final String actual = notificationService.send(createNotificationDto);

        assertThat(actual).isEqualTo("알림 전송 성공");
    }
});
```

이렇게 한 번 해보시면 좋을 것 같습니다

Reply:
추가적으로 통합 테스트에서는 `mocking`하는 과정을 최대한 지양했으면 좋겠습니다
외부 서비스는 별도로 두고 말이죠 
사실 외부 서비스는 아예 별도로 테스트용 `Stub`이나 `Fake` 등으로 작업하면 좋기는 하겠는데 이건 추후로 리펙토링을 하던가 팀단위로 정해야 할 것 같고..아무튼 그렇습니다